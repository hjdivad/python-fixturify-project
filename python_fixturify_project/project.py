import tempfile

from pathlib import Path

def deep_merge(a, b, path=None):
    "merges b into a"
    if path is None: path = []
    for key in b:
        if key in a:
            if isinstance(a[key], dict) and isinstance(b[key], dict):
                deep_merge(a[key], b[key], path + [str(key)])
            elif a[key] == b[key]:
                pass # same leaf value
            else:
                raise Exception('Conflict at %s' % '.'.join(path + [str(key)]))
        else:
            a[key] = b[key]
    return a

class Project(object):
    def __init__(self):
        self._base_dir = None
        self._tmp = None
        self._files = {}

    @property
    def base_dir(self):
        """Gets the base directory path, usually a tmp directory unless a baseDir has been explicitly set."""
        return self._base_dir

    @base_dir.setter
    def base_dir(self, value):
        """Sets the base directory of the project."""
        self._base_dir = value

    @base_dir.deleter
    def base_dir(self):
        """Deletes the base directory."""
        del self._base_dir

    @property
    def files(self):
        """Gets the files containing the project's directory structure."""
        return self._files

    @files.setter
    def files(self, value):
        """Sets the files corresponding to the project's directory structure."""
        self._files = value

    @files.deleter
    def files(self):
        """Deletes the files corresponding to the project's directory structure."""
        del self._files


    def __auto_base_dir(self):
        """"""
        if not self.base_dir:
            self.base_dir = tempfile.mkdtemp()

        return self.base_dir

    def merge_files(self, dir_json):
        """Merges an object containing a directory represention with the existing files."""
        self.files = deep_merge(self.files, dir_json)

    def write(self, dir_json):
        """Writes the existing files property containing a directory representation to the tmp directory."""
        if dir_json:
            self.merge_files(dir_json)

        self.write_project()

    def write_project(self):
        self.__auto_base_dir()


    def __write(self):
        Path(self.base_dir).mkdir(parents=True, exist_ok=True)

