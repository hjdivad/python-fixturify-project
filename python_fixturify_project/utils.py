from typing import Any, Dict, List, Tuple, Union

def deep_merge(a: Dict[Any, Any], b: Dict[Any, Any], path: List[str] = []) -> Dict[Any, Any]:
    """
    Merges dictionary b into dictionary a, with b overriding a for primitive values.
    """
    for key in b:
        if key in a:
            if isinstance(a[key], dict) and isinstance(b[key], dict):
                deep_merge(a[key], b[key], path + [str(key)])
            elif isinstance(a[key], (int, float, str, bool)) and isinstance(b[key], (int, float, str, bool)):
                a[key] = b[key]
            else:
                raise Exception("Conflict at %s" % ".".join(path + [str(key)]))
        else:
            a[key] = b[key]
    return a

def keys_exists(element: Dict[Any, Any], *keys: Union[str, int]) -> bool:
    """
    Check if *keys (nested) exists in `element` (dict).
    """
    if not isinstance(element, dict):
        raise AttributeError("keys_exists() expects dict as first argument.")
    if len(keys) == 0:
        raise AttributeError("keys_exists() expects at least two arguments, one given.")

    _element = element
    for key in keys:
        try:
            _element = _element[key]
        except KeyError:
            return False
    return True
