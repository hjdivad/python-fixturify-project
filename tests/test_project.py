"""Tests for Project class."""
from os import path
from pathlib import Path

import pytest
from conftest import BAD_DIR_NAME, BAD_EMPTY_NAME, GOOD_NESTED_DIRS, GOOD_SINGLE_FILE

from python_fixturify_project.exceptions import InvalidProjectError
from python_fixturify_project.project import Project


def test_project():
    """Example test with parametrization."""
    project = Project()
    assert isinstance(project, Project)


def test_cleanup_dir():
    # Given
    project = Project()
    project.write(GOOD_SINGLE_FILE)
    base_dir = project.base_dir

    assert path.exists(base_dir) and path.isdir(base_dir)

    # When
    del project

    # Then
    assert not path.exists(base_dir)


@pytest.mark.parametrize("test_input", [BAD_DIR_NAME, BAD_EMPTY_NAME])
def test_improper_write(test_input):
    with pytest.raises(InvalidProjectError):
        project = Project()
        project.write(test_input)


def test_proper_write_with_cleanup():
    base_dir = None

    with Project() as p:
        p.write(GOOD_NESTED_DIRS)
        base_dir = p.base_dir

        assert path.exists(base_dir) and path.isdir(base_dir)

        # Do some spot checking on the file structure
        file_checks = [
            (Path(base_dir, "valid_file.txt"), "file", "some text"),
            (Path(base_dir, "nested_dir"), "dir", None),
            (Path(base_dir, "nested_dir", "valid_empty_file"), "file", ""),
            (
                Path(
                    base_dir,
                    "nested_dir",
                    "another_nested_dir",
                    "last_nested_empty_dir",
                ),
                "dir",
                None,
            ),
            (
                Path(base_dir, "nested_dir", "another_nested_dir", "final_text_file"),
                "file",
                "some text",
            ),
        ]

        for file_path, file_type, file_contents in file_checks:
            if file_type == "file":
                assert path.exists(file_path) and path.isfile(file_path)
                with open(file_path) as _temp_file:
                    assert _temp_file.read() == file_contents
            else:
                assert path.exists(file_path) and path.isdir(file_path)

    # Ensure cleanup happened
    assert not path.exists(base_dir)
