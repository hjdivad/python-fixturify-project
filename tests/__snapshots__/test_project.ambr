# name: test_proper_write_with_cleanup
  dict({
    'nested_dir': dict({
      'another_nested_dir': dict({
        'final_text_file.txt': 'some text',
        'last_nested_empty_dir': dict({
        }),
      }),
      'another_nested_empty_dir': dict({
      }),
      'valid_empty_file.txt': '',
    }),
    'valid_file.txt': 'some text',
  })
# ---
# name: test_read
  dict({
    'nested_dir': dict({
      'another_nested_dir': dict({
        'final_text_file.txt': 'some text',
        'last_nested_empty_dir': dict({
        }),
      }),
      'another_nested_empty_dir': dict({
      }),
      'valid_empty_file.txt': '',
    }),
    'valid_file.txt': 'some text',
  })
# ---
